libbpf-rs/src/btf/mod.rs

```diff
@@ -0,0 +1,149 @@
+//! Parse and introspect btf information, from files or loaded objects.
+//!
+//! To find a specific type you can use one of 3 methods
+//!
+//! - [Btf::type_by_name]
+//! - [Btf::type_by_id]
+//! - [Btf::type_by_kind]
+//!
+//! All of these are generic over `K`, which is any type that can be created from a [`BtfType`],
+//! for all of these methods, not finding any type by the passed parameter or finding a type of
+//! another [`BtfKind`] will result in a [`None`] being returned (or filtered out in the case of
+//! [`Btf::type_by_kind`]). If you want to get a type independently of the kind, just make sure `K`
+//! binds to [`BtfType`].
+
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::marker::PhantomData;
+use std::mem::size_of;
+use std::os::raw::c_void;
+use std::os::unix::prelude::AsRawFd;
+use std::os::unix::prelude::FromRawFd;
+use std::os::unix::prelude::OsStrExt;
+use std::os::unix::prelude::OwnedFd;
+use std::path::Path;
+use std::ptr::NonNull;
+
+use crate::libbpf_sys;
+use crate::util::create_bpf_entity_checked;
+use crate::util::create_bpf_entity_checked_opt;
+use crate::util::parse_ret_i32;
+use crate::Error;
+use crate::Result;
+
+/// The btf information of a bpf object.
+///
+/// The lifetime bound protects against this object outliving its source. This can happen when it
+/// was derived from an [`Object`](super::Object), which owns the data this structs points too. When
+/// instead the [`Btf::from_path`] method is used, the lifetime will be `'static` since it doesn't
+/// borrow from anything.
+#[derive(Debug)]
+pub struct Btf<'source> {
+    ptr: NonNull<libbpf_sys::btf>,
+    needs_drop: bool,
+    _marker: PhantomData<&'source ()>,
+}
+
+impl Btf<'static> {
+    /// Load the btf information from an ELF file.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let path = {
+            let mut v = path.as_ref().as_os_str().as_bytes().to_vec();
+            v.push(0);
+            CString::from_vec_with_nul(v).map_err(|_| {
+                Error::InvalidInput(format!("invalid path {:?}, has null bytes", path.as_ref()))
+            })?
```

Comment:
Although this path to convert it is really roundabout, it's the only way that doesn't go through an utf8 check. Which technically makes this capable of accepting more paths.

Reply:
Sorry, I just came across this again. Doesn't `CString` have a constructor that creates an object from a `Vec`? https://doc.rust-lang.org/std/ffi/struct.CString.html#method.new

I would think something like this would work:
```rust
            CString::new(path.as_ref().as_os_str().as_bytes().to_vec())
```
?
This explicit NUL byte push is really putting me off right now and I see it even being conditional elsewhere...

Reply:
yeah, I did totally miss this obvious route.... we should change this to that version you just posted

Reply:
Should I make a PR or do you want to just push that small change to master directly?

Reply:
Feel free to go ahead and create a PR. Will prioritize reviewing it.